### ISSUE TITLE
(core): `crossRegionReferences` generates unnecessarily long SSM parameter names leading to deployment failures

### ISSUE THREAD

**Issue by colifran** [aws/aws-cdk#30119]
### Describe the bug

When using `crossRegionReferences` the producing stack creates SSM Parameters in the consuming region for each exported value. This is accomplished via this code https://github.com/aws/aws-cdk/blob/2779b5e7b4c962409f977eff2d28b96bc01c6c46/packages/aws-cdk-lib/core/lib/private/refs.ts#L248-L263. For users using longer stack names or nested stacks, the generated parameter name can become unnecessarily long. As a result, this can lead to deployment failures resulting from a response object that exceeds the [4096 byte](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-responses.html) limit when the export reader and writer custom resource providers are responding back to CloudFormation.

### Expected Behavior

`crossRegionReferences` doesn't generate unnecessarily long SSM parameter names.

### Current Behavior

`crossRegionReferences` generates SSM parameters with the name `/cdk/exports/{consumingStackName}/{export-name}`. For users using nested stacks or for users with longer stack names (or both), the parameter names can become unnecessarily long which can result in deployment failures.

### Reproduction Steps

```ts
import * as awscdk from 'aws-cdk-lib'

const ACCOUNT_ID = '111222333444'
const STACK_PREFIX = 'testing-cdk-cross-region-references-as-example-for-github-issue'
// 8 or less works, 9 or more fails with "Response object is too long"
const NUMBER_OF_KEYS_TO_CREATE = 9

function createRootStack (app: awscdk.App, region: string): awscdk.Stack {
  const stackName = `${STACK_PREFIX}-rootstack-${region}`

  const props = {
    env: {
      account: ACCOUNT_ID,
      region: region
    },
    crossRegionReferences: true,
    synthesizer: new awscdk.LegacyStackSynthesizer()
  }

  const stack = new awscdk.Stack(app, stackName, props)

  return stack
}

function createPrimaryKeyStack (rootStack: awscdk.Stack, keyName: string): string {
  const nestedStack = new awscdk.NestedStack(rootStack, `${STACK_PREFIX}-nestedstack-${keyName}`)

  const keyProps = getKeyProps(keyName)
  const kmsCfnKey = new awscdk.aws_kms.CfnKey(nestedStack, `${STACK_PREFIX}-key-${keyName}`, keyProps)

  const aliasName = `alias/${STACK_PREFIX}-${keyName}`
  const aliasProps = getAliasProps(aliasName, kmsCfnKey.attrKeyId)
  new awscdk.aws_kms.CfnAlias(nestedStack, `${STACK_PREFIX}-alias-${keyName}`, aliasProps)

  return kmsCfnKey.attrArn
}

function createReplicaKeyStack (rootStack: awscdk.Stack, keyName: string, primaryKeyArn: string) {
  const nestedStack = new awscdk.NestedStack(rootStack, `${STACK_PREFIX}-nestedstack-${keyName}`)

  const keyProps = getReplicaKeyProps(keyName, primaryKeyArn)
  const kmsCfnKey = new awscdk.aws_kms.CfnReplicaKey(nestedStack, `${STACK_PREFIX}-key-${keyName}`, keyProps)

  const aliasName = `alias/${STACK_PREFIX}-${keyName}`
  const aliasProps = getAliasProps(aliasName, kmsCfnKey.attrKeyId)
  new awscdk.aws_kms.CfnAlias(nestedStack, `${STACK_PREFIX}-alias-${keyName}`, aliasProps)
}

function getKeyPolicy (): awscdk.aws_iam.PolicyDocument {
  const thisPrincipal = new awscdk.aws_iam.AccountPrincipal(ACCOUNT_ID)

  const policyStatement = new awscdk.aws_iam.PolicyStatement({
    sid: 'Allow all key access',
    effect: awscdk.aws_iam.Effect.ALLOW,
    actions: [
      'kms:*'
    ],
    resources: ['*'],
    principals: [thisPrincipal]
  })

  const policyDocument = new awscdk.aws_iam.PolicyDocument({ statements: [policyStatement] })

  return policyDocument
}

function getKeyProps (keyName: string): awscdk.aws_kms.CfnKeyProps {
  const policyDocument = getKeyPolicy()

  const keyProps: awscdk.aws_kms.CfnKeyProps = {
    description: `Testing cross-region references in CDK - ${keyName}`,
    keyPolicy: policyDocument,
    multiRegion: true,
    enableKeyRotation: false,
    enabled: true,
    keyUsage: 'ENCRYPT_DECRYPT',
    pendingWindowInDays: 7
  }

  return keyProps
}

function getReplicaKeyProps (keyName: string, primaryKeyArn: string): awscdk.aws_kms.CfnReplicaKeyProps {
  const policyDocument = getKeyPolicy()

  const keyProps: awscdk.aws_kms.CfnReplicaKeyProps = {
    description: `Testing cross-region references in CDK - ${keyName}`,
    primaryKeyArn: primaryKeyArn,
    keyPolicy: policyDocument,
    enabled: true,
    pendingWindowInDays: 7
  }

  return keyProps
}

function getAliasProps (aliasName:string, targetKeyId: string): awscdk.aws_kms.CfnAliasProps {
  const aliasProps: awscdk.aws_kms.CfnAliasProps = {
    aliasName: aliasName,
    targetKeyId: targetKeyId
  }

  return aliasProps
}

function main (args: Array<string> | undefined = undefined) {
  const app = new awscdk.App()

  const rootStackEast = createRootStack(app, 'us-east-1')
  const rootStackWest = createRootStack(app, 'us-west-2')

  for (let i = 1; i <= NUMBER_OF_KEYS_TO_CREATE; i++) {
    const primaryKeyArn = createPrimaryKeyStack(rootStackEast, i.toString())
    createReplicaKeyStack(rootStackWest, i.toString(), primaryKeyArn)
  }
}

if (require.main === module) {
  main()
}
```

### Possible Solution

`crossRegionReferences` is an experimental feature. We may want to consider updating the SSM parameter names being generated to be limited to a certain length. This would break existing customers, though. Ideally, we could find a way to give users a choice to optionally limit the parameter name length. While this doesn't help us get around the 4096 byte limitation for custom resource provider response objects, it _would_ reduce the chances of a deployment failure by limiting the response body size.

### CDK CLI Version

2.136.0

### Node.js Version

20.11.0

### OS

MacOS

### Language

TypeScript

**Comment by renschler** [comment]
Thanks @colifran for reference I don't use nested stacks, and my stack prefixes are only 4 letters long, and I was still running into this issue due do the number of regions AWS has available.

**Comment by AlphaGarden** [comment]
Hi there, we are seeing the similar issue and wondering if there is any potential fix we can apply for fixing this response too long issue? Thanks in advance.

**Comment by IllarionovDimitri** [comment]
the other problem with created ssm parameter is that by default it will not be removed if the stack was deleted. For cloudfront i must deploy a global WAF in us-east-1 and reference it in other region where my other application stacks are deployed. I deployed once, remove all stacks and by re-deployment got following error `Resource handler returned message: "AWS WAF couldn?t perform the operation because some resource in your request is a duplicate of an existing one.` . It turned out that the ssm parameter from the very first deployment was still there. After i've removed it manually, i could deploy the stack again. Make sense to apply RemovalPolicy.DESTROY by default to the ssm paramter

**Comment by jcoetzee** [comment]
Couldn't this be unblocked (even as an interim solution) by creating a reader per reference instead of one reader for all references?

**Comment by jcoetzee** [comment]
If anyone is looking for a workaround that is actually useful (instead of using shorter stack names - chances are people are going to hit this when updating an existing stack). I wrote an Aspect that splits up the single`Custom::CrossRegionExportReader` resource that CDK defines into a reader per import. This completely bypasses the 4096 character problem (unless you have a single import that's that large, if that's even possible). The existing CDK implementation already defines a provider for the reader (which is unchanged) so this won't result in additional Lambda functions just separate invocations.

IMO updating to shorten the SSM parameter names is not the correct solution as it only delays the problem.

```
/* eslint-disable @typescript-eslint/no-explicit-any */
import { IAspect, Stack } from 'aws-cdk-lib';
import { IConstruct } from 'constructs';

/**
 * CDK Aspect that automatically splits Custom::CrossRegionExportReader resources
 * into one reader per import for every stack with `_crossRegionReferences = true`.
 *
 * This avoids the CloudFormation "Response object is too long" error (4096 byte limit)
 * that occurs when the batched reader returns too many exports in a single response.
 *
 * Usage (apply once at the App level):
 *   cdk.Aspects.of(app).add(new SplitCrossRegionExportReaders());
 */
export class SplitCrossRegionExportReaders implements IAspect {
  private readonly patched = new Set<Stack>();

  public visit(node: IConstruct): void {
    if (!(node instanceof Stack)) return;
    if (!(node as any)._crossRegionReferences) return;
    if (this.patched.has(node)) return;
    this.patched.add(node);
    splitCrossRegionReaders(node);
  }
}

/**
 * Splits any Custom::CrossRegionExportReader resources in a stack's template
 * into one resource per import. Can also be called directly on individual stacks.
 */
export function splitCrossRegionReaders(stack: Stack): void {
  const original = (stack as any)._toCloudFormation.bind(stack);
  (stack as any)._toCloudFormation = () => {
    const template = original();
    return transformTemplate(template);
  };
}

function transformTemplate(template: any): any {
  const resources = template.Resources;
  if (!resources) return template;

  const readersToSplit: string[] = [];

  for (const [logicalId, resource] of Object.entries<any>(resources)) {
    if (
      resource.Type === 'Custom::CrossRegionExportReader' &&
      resource.Properties?.ReaderProps?.imports &&
      Object.keys(resource.Properties.ReaderProps.imports).length > 1
    ) {
      readersToSplit.push(logicalId);
    }
  }

  for (const logicalId of readersToSplit) {
    splitReader(template, logicalId);
  }

  return template;
}

function splitReader(template: any, originalLogicalId: string): void {
  const resources = template.Resources;
  const original = resources[originalLogicalId];
  const readerProps = original.Properties.ReaderProps;
  const imports: Record<string, any> = readerProps.imports;
  const importKeys = Object.keys(imports);

  const newLogicalIds: Record<string, string> = {};

  importKeys.forEach((key, index) => {
    const newLogicalId = `${originalLogicalId}${index}`;
    newLogicalIds[key] = newLogicalId;

    resources[newLogicalId] = {
      Type: original.Type,
      Properties: {
        ServiceToken: original.Properties.ServiceToken,
        ReaderProps: {
          region: readerProps.region,
          prefix: readerProps.prefix,
          imports: { [key]: imports[key] },
        },
      },
      ...(original.DependsOn ? { DependsOn: original.DependsOn } : {}),
      ...(original.UpdateReplacePolicy ? { UpdateReplacePolicy: original.UpdateReplacePolicy } : {}),
      ...(original.DeletionPolicy ? { DeletionPolicy: original.DeletionPolicy } : {}),
    };
  });

  delete resources[originalLogicalId];

  rewriteRefs(template, originalLogicalId, newLogicalIds);
}

function rewriteRefs(obj: any, originalLogicalId: string, newLogicalIds: Record<string, string>): void {
  if (obj == null || typeof obj !== 'object') return;

  if (Array.isArray(obj)) {
    for (let i = 0; i < obj.length; i++) {
      obj[i] = maybeRewriteRef(obj[i], originalLogicalId, newLogicalIds);
      rewriteRefs(obj[i], originalLogicalId, newLogicalIds);
    }
    return;
  }

  for (const key of Object.keys(obj)) {
    obj[key] = maybeRewriteRef(obj[key], originalLogicalId, newLogicalIds);
    rewriteRefs(obj[key], originalLogicalId, newLogicalIds);
  }
}

function maybeRewriteRef(value: any, originalLogicalId: string, newLogicalIds: Record<string, string>): any {
  if (value == null || typeof value !== 'object') return value;

  const getAtt = value['Fn::GetAtt'];
  if (Array.isArray(getAtt) && getAtt.length === 2 && getAtt[0] === originalLogicalId) {
    const exportKey = getAtt[1];
    const newLogicalId = newLogicalIds[exportKey];
    if (newLogicalId) {
      return { 'Fn::GetAtt': [newLogicalId, exportKey] };
    }
  }

  if (value['Ref'] === originalLogicalId) {
    const firstKey = Object.keys(newLogicalIds)[0];
    if (firstKey) {
      return { Ref: newLogicalIds[firstKey] };
    }
  }

  if (Array.isArray(value) && value.includes(originalLogicalId)) {
    const idx = value.indexOf(originalLogicalId);
    value.splice(idx, 1, ...Object.values(newLogicalIds));
  }

  return value;
}
```
